File: examples/integration/low_memory_embed.c
#include "mosaic.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void free_ids(uint32_t *ids) {
    mosaic_free(ids);
}

int main(void) {
    const char *model = "fixtures/packs/model-v2.mpack";
    const char *unicode = "fixtures/packs/unicode17-v1.mpack";
    const char *language_en = "fixtures/packs/language/en-v1.mpack";
    const char *language_hi = "fixtures/packs/language/hi-v1.mpack";
    const char *language_ja = "fixtures/packs/language/ja-v1.mpack";
    const char *detector = "fixtures/packs/detector/reference-v1.mpack";
    const char *input = "tokenizer नमस्ते दुनिया こんにちは世界";
    mosaic_tokenizer *tokenizer = NULL;
    uint32_t *ids = NULL;
    size_t id_count = 0;
    mosaic_detection detection = {0};
    mosaic_status status = mosaic_tokenizer_load_files(model, unicode, &tokenizer);
    if (status != MOSAIC_OK) return 1;
    if ((status = mosaic_tokenizer_add_language_file(tokenizer, language_en)) != MOSAIC_OK) return 1;
    if ((status = mosaic_tokenizer_add_language_file(tokenizer, language_hi)) != MOSAIC_OK) return 1;
    if ((status = mosaic_tokenizer_add_language_file(tokenizer, language_ja)) != MOSAIC_OK) return 1;
    if ((status = mosaic_tokenizer_set_detector_file(tokenizer, detector)) != MOSAIC_OK) return 1;

    mosaic_runtime_limits limits;
    mosaic_runtime_limits_low_memory_default(&limits);
    if ((status = mosaic_tokenizer_set_runtime_limits(tokenizer, &limits)) != MOSAIC_OK) return 1;
    if ((status = mosaic_tokenizer_seal(tokenizer)) != MOSAIC_OK) return 1;

    status = mosaic_tokenizer_encode_auto(tokenizer, (const uint8_t *)input, strlen(input), &ids, &id_count, &detection);
    if (status != MOSAIC_OK) return 1;

    printf("route=%s tokens=%zu\n", detection.matched ? detection.language : "none", id_count);
    free_ids(ids);
    mosaic_tokenizer_free(tokenizer);
    return 0;
}

File: examples/integration/low_memory_embed.py
from __future__ import annotations

from pathlib import Path

from mosaic import NORMALIZE_NFC, BatchExecutor, Tokenizer


ROOT = Path(__file__).resolve().parents[2]
PACK = ROOT / "fixtures" / "packs"
LIB = ROOT / "build" / "preset-core-release" / "native" / "mosaic.dll"


def main() -> int:
    tokenizer = Tokenizer(PACK / "model-v2.mpack", PACK / "unicode17-v1.mpack", library_path=LIB)
    tokenizer.add_language(PACK / "language" / "en-v1.mpack")
    tokenizer.add_language(PACK / "language" / "hi-v1.mpack")
    tokenizer.add_language(PACK / "language" / "ja-v1.mpack")
    tokenizer.set_detector(PACK / "detector" / "reference-v1.mpack")
    tokenizer.set_low_memory_limits()
    tokenizer.seal()

    sample = "tokenizer नमस्ते दुनिया こんにちは世界".encode("utf-8")
    ids, detection = tokenizer.encode_auto(sample)
    assert tokenizer.decode(ids) == sample

    stream = tokenizer.online_stream(max_pending_bytes=64 * 1024)
    consumed, committed = stream.push(sample[:8])
    assert consumed <= 8
    _ = committed
    committed += stream.finish()

    assert tokenizer.normalize(b"e\xcc\x81", NORMALIZE_NFC).data == "é".encode("utf-8")

    with BatchExecutor.low_memory(library_path=LIB) as executor:
        results = executor.encode(tokenizer, [b"hello", b"world"])
        assert len(results) == 2
        assert all(result.status == 0 for result in results)

    print(f"route={detection.language} ids={len(ids)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
